Realm.Shared/Dynamic/MetaRealmObject.cs

```diff
@@ -0,0 +1,241 @@
+                    else
+                        realmObjectOpsMethod = typeof(RealmObjectOps).GetMethod(nameof(RealmObjectOps.SetStringValue));
+                    break;
+                case Schema.PropertyType.Data:
+                    argumentType = typeof(byte[]);
+                    realmObjectOpsMethod = typeof(RealmObjectOps).GetMethod(nameof(RealmObjectOps.SetByteArrayValue));
+                    break;
+                case Schema.PropertyType.Date:
+                    argumentType = typeof(DateTimeOffset);
+                    if (property.IsNullable)
+                        realmObjectOpsMethod = typeof(RealmObjectOps).GetMethod(nameof(RealmObjectOps.SetNullableDateTimeOffsetValue));
+                    else
+                        realmObjectOpsMethod = typeof(RealmObjectOps).GetMethod(nameof(RealmObjectOps.SetDateTimeOffsetValue));
+                    break;
+                case Schema.PropertyType.Object:
+                    argumentType = typeof(RealmObject);
+                    arguments.Insert(0, Expression.Field(GetLimitedSelf(), RealmObjectRealmField));
+                    realmObjectOpsMethod = typeof(RealmObjectOps).GetMethod(nameof(RealmObjectOps.SetObjectValue)).MakeGenericMethod(typeof(RealmObject));
+                    break;
+            }
+
+            if (property.IsNullable && argumentType.IsValueType)
+            {
+                argumentType = typeof(Nullable<>).MakeGenericType(argumentType);
+            }
+
+            var valueExpression = value.Expression;
+            if (valueExpression.Type != argumentType)
+            {
+                valueExpression = Expression.Convert(valueExpression, argumentType);
+            }
+
+            arguments.Add(valueExpression);
+
+            var expression = Expression.Block(Expression.Call(realmObjectOpsMethod, arguments), Expression.Default(binder.ReturnType));
+
+            var argumentShouldBeDynamicRealmObject = BindingRestrictions.GetTypeRestriction(Expression, typeof(DynamicRealmObject));
+            var argumentShouldBeInTheSameRealm = BindingRestrictions.GetInstanceRestriction(Expression.Field(GetLimitedSelf(), RealmObjectRealmField), _realm);
+            return new DynamicMetaObject(expression, argumentShouldBeDynamicRealmObject.Merge(argumentShouldBeInTheSameRealm));
+        }
+
+        public override IEnumerable<string> GetDynamicMemberNames()
+        {
+            return _metadata.Schema.PropertyNames;
+        }
+
+        private static Expression WeakConstant<T>(T value) where T : class
+        {
+            var weakReference = new WeakReference(value);
+            var constant = Expression.Constant(weakReference);
+            var property = Expression.Property(constant, nameof(weakReference.Target));
+            return Expression.Convert(property, typeof(T));
+        }
+
+        private Expression GetLimitedSelf()
```

Comment:
Why is this necessary? Could we please have more comments explaining the rationale behind the code.